File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
n = int(sys.stdin.readline())
graph = [ list(map(int, sys.stdin.readline().split())) for _ in range(n) ]
visit = [ False for _ in range(n) ] 
min_value = sys.maxsize 

def backTracking(depth, idx): 
      global min_value
      if depth == n // 2: 
            power1, power2 = 0, 0
            for i in range(n):
                  for j in range(n):
                        if visit[i] and visit[j]: 
                              power1 += graph[i][j]
                        elif not visit[i] and not visit[j]: 
                              power2 += graph[i][j]
            min_value = min(min_value, abs(power1-power2)) 
            return

      for i in range(idx, n): 
            if not visit[i]:
                  visit[i] = True
                  backTracking(depth+1, i+1) 
                  visit[i] = False
backTracking(0, 0)
print(min_value)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.*;
import java.util.*;

public class Main {
	static int n;
	static int ans = Integer.MAX_VALUE;
	static int[][] level;
	static boolean[] vis;

	public static void main(String args[]) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		n = Integer.parseInt(br.readLine());
		level = new int[n][n];
		vis = new boolean[n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			for (int j = 0; j < n; j++) {
				level[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		backTracking(0, 0);
		System.out.println(ans);
	}

	static void backTracking(int depth, int idx) {
		if (depth == n / 2) {
			int t1 = 0, t2 = 0;
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					if (vis[i] && vis[j])
						t1 += level[i][j];
					else if (!vis[i] && !vis[j])
						t2 += level[i][j];
				}
			}
			ans = Math.min(ans, Math.abs(t1 - t2));
			return;
		}

		for (int i = idx; i < n; i++) {
			if (!vis[i]) {
				vis[i] = true;
				backTracking(depth + 1, i + 1);
				vis[i] = false;
			}
		}
	}
}